File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
import java.util.*;

class Solution {

    public List<String> letterCombinations(String digits) {

        List<String> result = new ArrayList<>();

        if(digits == null || digits.length() == 0)
            return result;

        String[] phone = {
            "", "", "abc", "def", "ghi",
            "jkl", "mno", "pqrs", "tuv", "wxyz"
        };

        backtrack(result, phone, digits, 0, "");

        return result;
    }

    private void backtrack(List<String> result, String[] phone, String digits, int index, String current){

        if(index == digits.length()){
            result.add(current);
            return;
        }

        String letters = phone[digits.charAt(index) - '0'];

        for(char c : letters.toCharArray()){
            backtrack(result, phone, digits, index + 1, current + c);
        }
    }
}

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str):
        if not digits:
            return []
        
        phone = {
            "2":"abc", "3":"def", "4":"ghi",
            "5":"jkl", "6":"mno",
            "7":"pqrs", "8":"tuv", "9":"wxyz"
        }
        
        result = []
        
        def backtrack(index, path):
            if index == len(digits):
                result.append(path)
                return
            
            letters = phone[digits[index]]
            
            for letter in letters:
                backtrack(index + 1, path + letter)
        
        backtrack(0, "")
        return result